logging/proto_conversions.go

```diff
@@ -0,0 +1,81 @@
+package logging
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.viam.com/utils/protoutils"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// FieldToProto converts a zap.Field to a *structpb.Struct.
+func FieldToProto(field zap.Field) (*structpb.Struct, error) {
+	// Zap encodes float64s with very large int64s. Proto conversions have some
+	// loss with very large int64s. float32s are also encoded with int64s, but
+	// the int64 encodings are not large enough to cause loss in conversion.
+	// See https://pkg.go.dev/google.golang.org/protobuf@v1.32.0/types/known/structpb#NewValue.
+	//
+	// Use a hacky combination of fmt and math to store float64s as strings.
+	if field.Type == zapcore.Float64Type {
+		field.String = fmt.Sprintf("%f", math.Float64frombits(uint64(field.Integer)))
```

Comment:
This works? I figured if you wanted to pass the float as a string, you'd also need to change the `field.Type` to a `String`.

Reply:
Ah I'm keeping the type as `Float64Type` so that on the other side in `FieldKeyAndValueFromProto` we can detect that we're dealing with a float64 and `strconv.ParseFloat` from the `String` field. If we made it a `String` type, we would not be able to tell the difference between a bunch of digits in a string and a float64 encoded to a string.

Reply:
Before I go down some rabbit whole about documenting/communicating this:
* Does this change any existing behavior today about what gets into the app database when a main part logs directly with `Infow(msg, "field1", field1, ...)`
* With this patch about modules logging -> main -> app, will one see a `Field` state in the app database with the float64 type and both the `String` and `Integer` value filled in?

Reply:
> Does this change any existing behavior today about what gets into the app database when a main part logs directly with Infow(msg, "field1", field1, ...)

When a main part logs directly: no from what I understand. It _would_ if we changed [this code](https://github.com/viamrobotics/rdk/blob/main/logging/net_appender.go#L134-L145).

> With this patch about modules logging -> main -> app, will one see a Field state in the app database with the float64 type and both the String and Integer value filled in?

Yes I think so; and with current FE code (which, from what I can tell, only examines the `String` field), I bet float64 fields will now get logged as their bit representation (they were previously not being logged at all in the "Logs" tab).

Reply:
> Yes I think so; and with current FE code (which, from what I can tell, only examines the String field), I bet float64 fields will now get logged as their bit representation (they were previously not being logged at all in the "Logs" tab).

Yeah, the FE doesn't really work with the fields (I didn't realize it looked at string). But I did want to be clear about what exists in the database. If the FE changes to include more stuff, they'll need to understand the possible state of documents in the database.

So we should communicate this interface contract change. A peek at the zap `Field` object:
```
type Field struct {
	Key       string
	Type      FieldType
	Integer   int64
	String    string
	Interface interface{}
}
```

Where `FieldType` has a [mess of values](https://github.com/uber-go/zap/blob/master/zapcore/field.go#L35-L99).

Can you at least describe in documentation the state combinations for strings + ints + float32 + float64 and give guidance for how the other end should parse them (e.g: prefer the `Integer` value over the `String` value, but presenting the `String` is fine)? And let @zaporter-work decide if that should be duplicated anywhere for the front-end.

Reply:
Discussed offline, but I was incorrect in my original comment you're referencing there: there should be no change to how fields are stored in the app database. Also filed https://viam.atlassian.net/browse/APP-4143.